### PR TITLE
AVS-151 , AVS-153 Move tickets endpoints to GraphQL

### PR DESCRIPTION
**Pull request by You-Hsuan**

### CODE REVIEW

File: Access/Access.swift [hunk 1]
```diff
@@ -9,5 +9,9 @@
 import Foundation
 
 public protocol Access {
+    func getMyTickets(pageSize: Int, completion: @escaping (Result<PaginatedObject<Ticket>, Error>) -> Void)
+    func getMyTicketById(id: Int, completion: @escaping (Result<Ticket, Error>) -> Void)
+    func getUpcomingTicket(completion: @escaping (Result<Ticket?, Error>) -> Void)
```
[the remark this answers is not in the document]
Comment (You-Hsuan) [reply Access/Access.swift:14]: @rossrossp from the original v3 implementation, we only get one upcoming ticket where pageSize is 1. Should we change that?

File: Access/AccessImplementing.swift [hunk 2]
```diff
@@ -19,6 +19,84 @@ public class AccessImplementing: Access {
...
+                        return completion(.failure(GraphQLManagerError.noDataError))
+                    }
+                    completion(.success(ticket))
+                case .failure(let error):
+                    completion(.failure(error))
+                }
+            }
+    }
+
+    public func getUpcomingTicket(completion: @escaping (Result<Ticket?, Error>) -> Void) {
```
[the remark this answers is not in the document]
Comment (You-Hsuan) [reply Access/AccessImplementing.swift:60]: @rossrossp from the original v3 implementation, we only get one upcoming ticket where pageSize set to 1. Should we change that?

File: Access/Models/Ticket/AdditionalField.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+//  Created by YOU-HSUAN YU on 2023/1/27.
+//  Copyright © 2023 Realife Tech. All rights reserved.
+//
+
+import Foundation
+#if !COCOAPODS
+import GraphQL
+#endif
+
+public struct AdditionalField: Codable, Equatable {
```
[the remark this answers is not in the document]
Comment (You-Hsuan) [reply Access/Models/Ticket/AdditionalField.swift:14]: That's a good point. I'll rename to TicketAdditionalField. Thank you!

File: Access/AccessImplementing.swift [hunk 4]
```diff
@@ -19,6 +19,84 @@ public class AccessImplementing: Access {
         self.graphQLManager = graphQLManager
     }
 
+    public func getMyTickets(pageSize: Int, completion: @escaping (Result<PaginatedObject<Ticket>, Error>) -> Void) {
+        graphQLManager.dispatch(
+            query: ApolloType.GetMyTicketsQuery(pageSize: pageSize),
+            cachePolicy: .fetchIgnoringCacheData
```
[the remark this answers is not in the document]
Comment (You-Hsuan) [reply Access/AccessImplementing.swift:25]: @rossrossp I think ticketing should only work when app is online as the ticket UI will change base on the ticket's status from server. For example, app will blur the ticket qrcode when that ticket has been redeemed from someone else. And also it's the same cache mechanism as v3, so I think we can keep the same.

[the remark this answers is not in the document]
Comment (You-Hsuan) [reply Access/AccessImplementing.swift:25]: Update: Had a meeting with Ross and Richard, we've decided to change the cache policy to `returnCacheDataAndFetch`

**Review by rossrossp — APPROVED** [review]

**Pull request merged**